File: src/trading_assistant/features/technical/indicators.py
from __future__ import annotations

import logging
from typing import Iterable

import pandas as pd

log = logging.getLogger(__name__)


def compute_indicators(frame: pd.DataFrame, indicators: Iterable[str]) -> pd.DataFrame:
    """Compute configured indicators using pandas-ta; NaNs remain for warm-up rows."""
    try:
        import pandas_ta as ta
    except ImportError as exc:
        raise RuntimeError("pandas-ta is required for technical feature computation") from exc

    df = frame.copy()
    df.columns = [str(c).lower() for c in df.columns]
    required = {"open", "high", "low", "close", "volume"}
    missing = required - set(df.columns)
    if missing:
        raise ValueError(f"OHLCV frame missing columns: {sorted(missing)}")
    wanted = set(indicators)
    if "sma" in wanted:
        df["sma_20"] = ta.sma(df["close"], length=20)
        df["sma_50"] = ta.sma(df["close"], length=50)
        df["sma_200"] = ta.sma(df["close"], length=200)
    if "ema" in wanted:
        df["ema_12"] = ta.ema(df["close"], length=12)
        df["ema_26"] = ta.ema(df["close"], length=26)
    if "macd" in wanted:
        macd = ta.macd(df["close"], fast=12, slow=26, signal=9)
        if macd is not None:
            df["macd"] = macd.iloc[:, 0]
            df["macd_histogram"] = macd.iloc[:, 1]
            df["macd_signal"] = macd.iloc[:, 2]
    if "rsi" in wanted:
        df["rsi_14"] = ta.rsi(df["close"], length=14)
    if "stoch" in wanted:
        stoch = ta.stoch(df["high"], df["low"], df["close"], k=14, d=3, smooth_k=3)
        if stoch is not None:
            df["stoch_k"] = stoch.iloc[:, 0]
            df["stoch_d"] = stoch.iloc[:, 1]
    if "bollinger" in wanted:
        bands = ta.bbands(df["close"], length=20, std=2)
        if bands is not None:
            df["bb_lower"] = bands.iloc[:, 0]
            df["bb_middle"] = bands.iloc[:, 1]
            df["bb_upper"] = bands.iloc[:, 2]
    if "atr" in wanted:
        df["atr_14"] = ta.atr(df["high"], df["low"], df["close"], length=14)
    if "obv" in wanted:
        df["obv"] = ta.obv(df["close"], df["volume"])
    if "volume_sma" in wanted:
        df["volume_sma_20"] = ta.sma(df["volume"], length=20)
    if "adx" in wanted:
        adx = ta.adx(df["high"], df["low"], df["close"], length=14)
        if adx is not None:
            df["adx_14"] = adx["ADX_14"]
            df["plus_di_14"] = adx["DMP_14"]
            df["minus_di_14"] = adx["DMN_14"]
    if "ichimoku" in wanted:
        ichimoku = ta.ichimoku(df["high"], df["low"], df["close"], tenkan=9, kijun=26, senkou=52)
        if ichimoku is not None:
            # ta.ichimoku returns (historical, forward-projected) - only the
            # first, index-aligned frame is usable as a per-bar feature; the
            # second frame projects the cloud into future dates that don't
            # exist yet in this index and would misalign on merge.
            historical = ichimoku[0]
            df["ichimoku_tenkan"] = historical["ITS_9"]
            df["ichimoku_kijun"] = historical["IKS_26"]
            df["ichimoku_senkou_a"] = historical["ISA_9"]
            df["ichimoku_senkou_b"] = historical["ISB_26"]
            df["ichimoku_chikou"] = historical["ICS_26"]
    if "candles" in wanted:
        df = _compute_candle_patterns(df)
    if "volatility_divergence" in wanted:
        returns = df["close"].pct_change()
        df["volatility_20"] = returns.rolling(20, min_periods=5).std()
        # Autocorrelation of daily returns at lag 1 over a trailing 20-bar
        # window: positive means recent moves tend to persist (momentum),
        # negative means they tend to reverse (mean reversion), computed
        # causally (rolling, backward-looking only).
        df["return_autocorr_20"] = returns.rolling(20, min_periods=10).apply(
            lambda window: window.autocorr(lag=1), raw=False
        )
        # Volume/price divergence: correlation between price direction and
        # volume over a trailing window. A strong uptrend on falling volume
        # (low/negative divergence) is a classic "weak hands" warning signal
        # technical analysts watch for; here it's a continuous feature
        # rather than a hand-tuned rule.
        volume_change = df["volume"].pct_change()
        df["volume_price_divergence"] = returns.rolling(20, min_periods=10).corr(volume_change)
    return df


def _compute_candle_patterns(df: pd.DataFrame) -> pd.DataFrame:
    """Hand-rolled single-bar candlestick patterns.

    pandas-ta's `cdl_pattern` delegates to the compiled TA-Lib C library for
    most named patterns, which is not part of this project's pure-Python
    dependency stack (see pyproject.toml) and is nontrivial to build in
    minimal/CI environments. These five are simple enough to define directly
    from OHLC ratios with no external dependency, and the definitions are
    standard textbook ones - not an attempt to reproduce TA-Lib exactly.
    """
    body = (df["close"] - df["open"]).abs()
    candle_range = (df["high"] - df["low"]).replace(0, pd.NA)
    upper_wick = df["high"] - df[["open", "close"]].max(axis=1)
    lower_wick = df[["open", "close"]].min(axis=1) - df["low"]

    # candle_body_ratio is meant to be a fraction of the bar's total range in
    # [0, 1]. Real OHLC feeds occasionally contain a bar where open or close
    # falls outside [low, high] (bad ticks, split/dividend adjustment
    # artifacts, etc.), which would otherwise let body > candle_range and
    # push the ratio above 1 - clip defensively rather than let a single bad
    # bar produce a nonsensical feature value silently.
    df["candle_body_ratio"] = (body / candle_range).clip(upper=1.0)
    # Doji: body is a small fraction of the bar's total range - open and
    # close are nearly equal, signaling indecision.
    df["candle_doji"] = (df["candle_body_ratio"] < 0.1).astype(int)
    # Hammer: small body near the top of the range with a long lower wick
    # (>= 2x the body) and a short upper wick - a classic potential-reversal
    # shape after a decline.
    df["candle_hammer"] = (
        (lower_wick >= 2 * body) & (upper_wick <= body) & (body > 0)
    ).astype(int)

    prev_open, prev_close = df["open"].shift(1), df["close"].shift(1)
    prev_body_low = pd.concat([prev_open, prev_close], axis=1).min(axis=1)
    prev_body_high = pd.concat([prev_open, prev_close], axis=1).max(axis=1)
    # Bullish engulfing: prior bar closed down, current bar closes up and its
    # body fully engulfs the prior bar's body.
    df["candle_bullish_engulfing"] = (
        (prev_close < prev_open) & (df["close"] > df["open"]) &
        (df["open"] <= prev_body_low) & (df["close"] >= prev_body_high)
    ).astype(int)
    # Bearish engulfing: the mirror image - prior bar closed up, current bar
    # closes down and fully engulfs the prior body.
    df["candle_bearish_engulfing"] = (
        (prev_close > prev_open) & (df["close"] < df["open"]) &
        (df["open"] >= prev_body_high) & (df["close"] <= prev_body_low)
    ).astype(int)
    return df

File: server/indicators.ts
import { PriceBar, TechnicalFeature } from './db.js';

export function calculateIndicators(bars: PriceBar[]): TechnicalFeature[] {
  if (!bars || bars.length === 0) return [];

  // Sort bars chronologically
  const sorted = [...bars].sort((a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime());

  const closes = sorted.map((b) => b.close);
  const highs = sorted.map((b) => b.high);
  const lows = sorted.map((b) => b.low);
  const opens = sorted.map((b) => b.open);
  const volumes = sorted.map((b) => b.volume || 0);

  const n = sorted.length;
  const results: TechnicalFeature[] = [];

  // Helper functions
  const sma = (series: number[], period: number, idx: number): number | null => {
    if (idx < period - 1) return null;
    let sum = 0;
    for (let i = idx - period + 1; i <= idx; i++) sum += series[i];
    return sum / period;
  };

  // EMA series calculation
  const calcEMA = (series: number[], period: number): (number | null)[] => {
    const emaArr: (number | null)[] = [];
    const k = 2 / (period + 1);
    let prevEma: number | null = null;
    for (let i = 0; i < series.length; i++) {
      if (i < period - 1) {
        emaArr.push(null);
      } else if (i === period - 1) {
        let sum = 0;
        for (let j = 0; j < period; j++) sum += series[j];
        prevEma = sum / period;
        emaArr.push(prevEma);
      } else {
        prevEma = series[i] * k + (prevEma as number) * (1 - k);
        emaArr.push(prevEma);
      }
    }
    return emaArr;
  };

  const ema12 = calcEMA(closes, 12);
  const ema26 = calcEMA(closes, 26);

  // MACD Line = EMA12 - EMA26
  const macdLine: (number | null)[] = [];
  for (let i = 0; i < n; i++) {
    if (ema12[i] !== null && ema26[i] !== null) {
      macdLine.push((ema12[i] as number) - (ema26[i] as number));
    } else {
      macdLine.push(null);
    }
  }

  // MACD Signal = 9 period EMA of MACD Line
  const validMacdIndices = macdLine.map((v, i) => (v !== null ? i : -1)).filter((i) => i !== -1);
  const macdSignal: (number | null)[] = new Array(n).fill(null);
  const macdHist: (number | null)[] = new Array(n).fill(null);

  if (validMacdIndices.length >= 9) {
    const validMacdVals = validMacdIndices.map((i) => macdLine[i] as number);
    const sigVals = calcEMA(validMacdVals, 9);
    for (let k = 0; k < validMacdIndices.length; k++) {
      const origIdx = validMacdIndices[k];
      macdSignal[origIdx] = sigVals[k];
      if (macdLine[origIdx] !== null && sigVals[k] !== null) {
        macdHist[origIdx] = (macdLine[origIdx] as number) - (sigVals[k] as number);
      }
    }
  }

  // RSI 14
  const rsi: (number | null)[] = new Array(n).fill(null);
  if (n > 14) {
    let gains = 0;
    let losses = 0;
    for (let i = 1; i <= 14; i++) {
      const diff = closes[i] - closes[i - 1];
      if (diff >= 0) gains += diff;
      else losses -= diff;
    }
    let avgGain = gains / 14;
    let avgLoss = losses / 14;
    rsi[14] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);

    for (let i = 15; i < n; i++) {
      const diff = closes[i] - closes[i - 1];
      const gain = diff > 0 ? diff : 0;
      const loss = diff < 0 ? -diff : 0;
      avgGain = (avgGain * 13 + gain) / 14;
      avgLoss = (avgLoss * 13 + loss) / 14;
      rsi[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
    }
  }

  // OBV
  let currentObv = 0;
  const obvArr: number[] = [0];
  for (let i = 1; i < n; i++) {
    if (closes[i] > closes[i - 1]) currentObv += volumes[i];
    else if (closes[i] < closes[i - 1]) currentObv -= volumes[i];
    obvArr.push(currentObv);
  }

  // Generate full features
  for (let i = 0; i < n; i++) {
    const bar = sorted[i];
    const s20 = sma(closes, 20, i);
    const s50 = sma(closes, 50, i);
    const s200 = sma(closes, 200, i);
    const volSma20 = sma(volumes, 20, i);

    // Bollinger bands (20, 2 std)
    let bbLower: number | null = null;
    let bbMiddle: number | null = s20;
    let bbUpper: number | null = null;
    if (i >= 19 && s20 !== null) {
      let variance = 0;
      for (let j = i - 19; j <= i; j++) variance += Math.pow(closes[j] - s20, 2);
      const std = Math.sqrt(variance / 20);
      bbLower = s20 - 2 * std;
      bbUpper = s20 + 2 * std;
    }

    // Stochastic %K %D (14, 3)
    let stochK: number | null = null;
    let stochD: number | null = null;
    if (i >= 13) {
      let minLow = lows[i];
      let maxHigh = highs[i];
      for (let j = i - 13; j <= i; j++) {
        if (lows[j] < minLow) minLow = lows[j];
        if (highs[j] > maxHigh) maxHigh = highs[j];
      }
      const range = maxHigh - minLow;
      stochK = range > 0 ? ((closes[i] - minLow) / range) * 100 : 50;
    }

    // ATR 14
    let atr14: number | null = null;
    if (i >= 13) {
      let trSum = 0;
      for (let j = i - 13; j <= i; j++) {
        const prevClose = j > 0 ? closes[j - 1] : opens[j];
        const tr = Math.max(highs[j] - lows[j], Math.abs(highs[j] - prevClose), Math.abs(lows[j] - prevClose));
        trSum += tr;
      }
      atr14 = trSum / 14;
    }

    // Candle patterns
    const body = Math.abs(closes[i] - opens[i]);
    const candleRange = Math.max(0.0001, highs[i] - lows[i]);
    const upperWick = highs[i] - Math.max(opens[i], closes[i]);
    const lowerWick = Math.min(opens[i], closes[i]) - lows[i];
    const bodyRatio = Math.min(1.0, body / candleRange);
    const candleDoji = bodyRatio < 0.1 ? 1 : 0;
    const candleHammer = lowerWick >= 2 * body && upperWick <= body && body > 0 ? 1 : 0;

    let candleBullishEngulfing = 0;
    let candleBearishEngulfing = 0;
    if (i > 0) {
      const prevOpen = opens[i - 1];
      const prevClose = closes[i - 1];
      const prevBodyLow = Math.min(prevOpen, prevClose);
      const prevBodyHigh = Math.max(prevOpen, prevClose);
      if (prevClose < prevOpen && closes[i] > opens[i] && opens[i] <= prevBodyLow && closes[i] >= prevBodyHigh) {
        candleBullishEngulfing = 1;
      }
      if (prevClose > prevOpen && closes[i] < opens[i] && opens[i] >= prevBodyHigh && closes[i] <= prevBodyLow) {
        candleBearishEngulfing = 1;
      }
    }

    // Returns
    const return1 = i >= 1 ? (closes[i] - closes[i - 1]) / closes[i - 1] : null;
    const return3 = i >= 3 ? (closes[i] - closes[i - 3]) / closes[i - 3] : null;
    const return5 = i >= 5 ? (closes[i] - closes[i - 5]) / closes[i - 5] : null;
    const return10 = i >= 10 ? (closes[i] - closes[i - 10]) / closes[i - 10] : null;

    // Volatility 20
    let vol20: number | null = null;
    if (i >= 19) {
      const rets: number[] = [];
      for (let j = i - 19; j <= i; j++) {
        if (j > 0) rets.push((closes[j] - closes[j - 1]) / closes[j - 1]);
      }
      const mean = rets.reduce((a, b) => a + b, 0) / rets.length;
      const variance = rets.reduce((a, b) => a + Math.pow(b - mean, 2), 0) / rets.length;
      vol20 = Math.sqrt(variance);
    }

    const date = new Date(bar.timestamp);

    results.push({
      id: bar.id,
      asset_id: bar.asset_id,
      interval: bar.interval,
      timestamp: bar.timestamp,
      sma_20: s20 ? parseFloat(s20.toFixed(4)) : null,
      sma_50: s50 ? parseFloat(s50.toFixed(4)) : null,
      sma_200: s200 ? parseFloat(s200.toFixed(4)) : null,
      ema_12: ema12[i] !== null ? parseFloat((ema12[i] as number).toFixed(4)) : null,
      ema_26: ema26[i] !== null ? parseFloat((ema26[i] as number).toFixed(4)) : null,
      macd: macdLine[i] !== null ? parseFloat((macdLine[i] as number).toFixed(4)) : null,
      macd_signal: macdSignal[i] !== null ? parseFloat((macdSignal[i] as number).toFixed(4)) : null,
      macd_histogram: macdHist[i] !== null ? parseFloat((macdHist[i] as number).toFixed(4)) : null,
      rsi_14: rsi[i] !== null ? parseFloat((rsi[i] as number).toFixed(2)) : null,
      stoch_k: stochK !== null ? parseFloat(stochK.toFixed(2)) : null,
      stoch_d: stochD !== null ? parseFloat(stochD.toFixed(2)) : null,
      bb_lower: bbLower !== null ? parseFloat(bbLower.toFixed(4)) : null,
      bb_middle: bbMiddle !== null ? parseFloat(bbMiddle.toFixed(4)) : null,
      bb_upper: bbUpper !== null ? parseFloat(bbUpper.toFixed(4)) : null,
      atr_14: atr14 !== null ? parseFloat(atr14.toFixed(4)) : null,
      obv: obvArr[i],
      volume_sma_20: volSma20 ? parseFloat(volSma20.toFixed(2)) : null,
      candle_body_ratio: parseFloat(bodyRatio.toFixed(4)),
      candle_doji: candleDoji,
      candle_hammer: candleHammer,
      candle_bullish_engulfing: candleBullishEngulfing,
      candle_bearish_engulfing: candleBearishEngulfing,
      return_1: return1 !== null ? parseFloat(return1.toFixed(5)) : null,
      return_3: return3 !== null ? parseFloat(return3.toFixed(5)) : null,
      return_5: return5 !== null ? parseFloat(return5.toFixed(5)) : null,
      return_10: return10 !== null ? parseFloat(return10.toFixed(5)) : null,
      volatility_20: vol20 !== null ? parseFloat(vol20.toFixed(5)) : null,
      day_of_week: date.getUTCDay(),
      dist_from_high: highs[i] > 0 ? parseFloat(((highs[i] - closes[i]) / highs[i]).toFixed(4)) : 0,
      dist_from_low: lows[i] > 0 ? parseFloat(((closes[i] - lows[i]) / lows[i]).toFixed(4)) : 0,
    });
  }

  return results;
}
